csharp/src/Ice/Communicator.cs

```diff
@@ -1887,75 +1890,83 @@ internal OutgoingConnectionFactory OutgoingConnectionFactory()
         // Return the C# class associated with this Slice type-id
         // Used for both non-local Slice classes and exceptions
         //
-        internal Type? ResolveClass(string id)
+        internal Type? ResolveClass(string typeId)
         {
-            // First attempt corresponds to no cs:namespace metadata in the
-            // enclosing top-level module
-            //
-            string className = TypeToClass(id);
-            Type? c = AssemblyUtil.FindType(className);
-
-            //
-            // If this fails, look for helper classes in the typeIdNamespaces namespace(s)
-            //
-            if (c == null && _typeIdNamespaces != null)
+            return _typeIdCache.GetOrAdd(typeId, id =>
             {
-                foreach (string ns in _typeIdNamespaces)
+                // First attempt corresponds to no cs:namespace metadata in the
+                // enclosing top-level module
+                string className = TypeToClass(id);
+                Type? classType = AssemblyUtil.FindType(className);
+
+                //
+                // If this fails, look for helper classes in the typeIdNamespaces namespace(s)
+                //
+                if (classType == null && _typeIdNamespaces != null)
```

Comment:
I think we can remove the `_typeIdNamespaces != null` check `_typeIdNamespaces` is non null reference